### PR TITLE
test: Adds error and migration tests

### PR DESCRIPTION
**Pull request by EspenAlbert**
## Description

Changes:
- Add migration tests
- Add and Support TestAccStreamProcessor_JSONWhiteSpaceFormat
  - See [Failed test](https://github.com/mongodb/terraform-provider-mongodbatlas/actions/runs/10504792980/job/29101199975?pr=2519)
 - Fix updates to `state` and other attribute simultaneously should raise error
   - See [Failed test](https://github.com/mongodb/terraform-provider-mongodbatlas/actions/runs/10505399331/job/29103069516)
- Fix exit/return when invalid state is set on create, see [e880821af2537f8de20cb05b349510738abc0aef](https://github.com/mongodb/terraform-provider-mongodbatlas/pull/2519/commits/e880821af2537f8de20cb05b349510738abc0aef)
- Fix invalid transition from CREATED -> STOPPED [1f06993](https://github.com/mongodb/terraform-provider-mongodbatlas/pull/2519/commits/1f06993b9bf1ff2045b83979a4c6263fe26bfbe2)

Link to any related issue(s): CLOUDP-269090

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [ ] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR. A migration guide must be created or updated if the new feature will go in a major version.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR. A migration guide must be created or updated.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have read the [contributing guides](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/contributing/README.md)
- [x] I have checked that this change does not generate any credentials and that **they are NOT accidentally logged anywhere**.
- [x] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run make fmt and formatted my code
- [x] If changes include deprecations or removals, I defined an isolated PR with a relevant title as it will be used in the auto-generated changelog.
- [x] If changes include removal or addition of 3rd party GitHub actions, I updated our internal document. Reach out to the APIx Integration slack channel to get access to the internal document.

### CODE REVIEW

File: internal/common/schemafunc/json.go [hunk 1]
```diff
@@ -1,42 +1,11 @@
...
 import (
-	"context"
 	"encoding/json"
 	"log"
 	"reflect"
-
-	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
 )
 
-func DiffSuppressJSON() planmodifier.String {
```
Comment (EspenAlbert) [inline internal/common/schemafunc/json.go:12]: This didn't work as expected. I was trying to add whitespace to the `pipeline` attribute, but still, an update was shown.
Worked when I instead used a `func (r *streamProcessorRS) ModifyPlan(ctx context.Context, req resource.ModifyPlanRequest, resp *resource.ModifyPlanResponse)` on the resource.
However, it is too hacky to include. Considering users adding additional whitespace to be a low chance.

File: internal/service/streamprocessor/resource_test.go [hunk 2]
```diff
@@ -128,25 +154,58 @@ func TestAccStreamProcessor_clusterType(t *testing.T) {
...
+		projectID         = acc.ProjectIDExecution(t)
+		processorName     = "new-processor"
+		instanceName      = acc.RandomName()
+		invalidJSONConfig = connectionConfig{connectionType: connTypeSample, pipelineStepIsSource: true, invalidJSON: true}
 	)
 
 	resource.ParallelTest(t, resource.TestCase{
 		PreCheck:                 func() { acc.PreCheckBasic(t) },
 		ProtoV6ProviderFactories: acc.TestAccProviderV6Factories,
-		CheckDestroy:             checkDestroyStreamProcessor,
```
Comment (EspenAlbert) [inline internal/service/streamprocessor/resource_test.go:141]: no stream processor is created

File: internal/service/streamprocessor/resource_test.go [hunk 3]
```diff
@@ -61,6 +68,25 @@ func TestAccStreamProcessor_basic(t *testing.T) {
...
+	)
+	resource.ParallelTest(t, resource.TestCase{
+		ProtoV6ProviderFactories: acc.TestAccProviderV6Factories,
+		PreCheck:                 func() { acc.PreCheckBasic(t) },
+		CheckDestroy:             checkDestroyStreamProcessor,
+		Steps: []resource.TestStep{
+			{
+				Config: config(t, projectID, instanceName, processorName, streamprocessor.CreatedState, sampleSrcConfigExtraSpaces, testLogDestConfig),
+				Check:  composeStreamProcessorChecks(projectID, instanceName, processorName, streamprocessor.CreatedState, false, false),
+			},
```
Comment (oarbusi) [inline internal/service/streamprocessor/resource_test.go:89]: I don't see the point of testing that a json string with whitespaces is valid. I think this is not provide any value in the test suite and can be distracting

File: internal/service/streamprocessor/model.go [hunk 4]
```diff
@@ -52,14 +53,21 @@ func NewStreamProcessorWithStats(ctx context.Context, projectID, instanceName st
 	if diags.HasError() {
 		return nil, diags
 	}
+	// we need to use pipelinePlan as the state must match the configuration in case the json has extra whitespace or different ordering
```
Comment (oarbusi) [inline internal/service/streamprocessor/model.go]: I don't fully understand the reasoning here. What cases are you thinking here?

Comment (EspenAlbert) [reply internal/service/streamprocessor/model.go]: If we use only the `apiResp.GetPipeline()` it means we are only accepting stream_processors with "standard" JSON format.
E.g., suppose the api returns:
```json
{
    // other attributes removed from exampel
    "pipeline": [{"$source":{"connectionName" :"sample_stream_solar"}},{"$emit":{"connectionName": "__testLog"}}] // notice that this is not a string
}
```
We will always dump this to:
`"[{\"$source\":{\"connectionName\":\"sample_stream_solar\"}},{\"$emit\":{\"connectionName\": \"__testLog\"}}]"`

This means that if the user has any other format that doesn't match our `dump` format they will get an error:
```
╷
│ Error: Provider produced inconsistent result after apply
│ 
│ When applying changes to mongodbatlas_stream_processor.processor, provider "provider[\"registry.terraform.io/mongodb/mongodbatlas\"]" produced an unexpected new value: .pipeline: was
│ cty.StringVal("[{\"$source\":{\"connectionName\" : \"sample_stream_solar\"}},{\"$emit\":{\"connectionName\": \"__testLog\"}}]"), but now
│ cty.StringVal("[{\"$source\":{\"connectionName\":\"sample_stream_solar\"}},{\"$emit\":{\"connectionName\":\"__testLog\"}}]").
│ 
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.
```
- This example shows only an extra whitespace
- But I can imagine usage of the `templatefile()` in the `*.tf` file reading a human-readable JSON and changing some values

WDYT? Luckily `jsonencode()` does the same as our `dump` function but isn't it convenient to be less strict? 🤔

File: internal/service/streamprocessor/resource_test.go (hunk 3 above)
Comment (EspenAlbert) [reply internal/service/streamprocessor/resource_test.go:89]: Please see [comment below for more explanation](https://github.com/mongodb/terraform-provider-mongodbatlas/pull/2519#discussion_r1727034943)

File: internal/common/fwtypes/json_string.go [hunk 5]
```diff
@@ -0,0 +1,141 @@
+package fwtypes
```
Comment (EspenAlbert) [inline internal/common/fwtypes/json_string.go:1]: This is to allow different JSON formats (whitespace/no-whitespace/line breaks etc.)
- See [this comment](https://github.com/mongodb/terraform-provider-mongodbatlas/pull/2519#discussion_r1727034943) for a full example.
- See [the TPF docs](https://developer.hashicorp.com/terraform/plugin/framework/migrating/resources/crud#planned-value-does-not-match-config-value) for the `Planned Value does not match Config Value` workaround.

\* More background why `StateFunc` was removed from SDK v2. Similar example for [`iam_role`](https://github.com/hashicorp/terraform-provider-aws/blob/e2afe00d6207f95b837fc35518efba1572da7908/internal/service/iam/role.go#L61)
> as of Terraform 0.12, it’s required that:
> Resources should never set or change an attribute value without the schema Computed flag.
> Resources should always set an attribute state value to the exact configuration value or prior state value, if not null.
> Both of these rules are being violated with that usage of StateFunc and are the main reason a similar functionality cannot be achieved with Plugin Framework 2. If you try to change a configured value in Plugin Framework you’ll get an error like:

File: internal/common/schemafunc/json.go (hunk 1 above)
Comment (EspenAlbert) [reply internal/common/schemafunc/json.go:12]: Not necessary. See above: https://github.com/mongodb/terraform-provider-mongodbatlas/pull/2519/files#r1728618273

File: internal/service/streamprocessor/resource_schema.go [hunk 6]
```diff
@@ -30,12 +28,7 @@ func ResourceSchema(ctx context.Context) schema.Schema {
 				MarkdownDescription: "Human-readable label that identifies the stream instance.",
 			},
 			"pipeline": schema.StringAttribute{
-				Validators: []validator.String{
-					validate.StringIsJSON(),
-				},
-				PlanModifiers: []planmodifier.String{
-					schemafunc.DiffSuppressJSON(),
-				},
+				CustomType:          fwtypes.JSONStringType,
```
Comment (EspenAlbert) [inline internal/service/streamprocessor/resource_schema.go:31]: Finally, a usage of `CustomType`!

File: internal/service/streamprocessor/resource_test.go (hunk 3 above)
Comment (EspenAlbert) [reply internal/service/streamprocessor/resource_test.go:89]: ok to resolve @oarbusi ?

File: internal/service/streamprocessor/model.go (hunk 4 above)
Comment (EspenAlbert) [reply internal/service/streamprocessor/model.go]: A bit outdated, see comment below https://github.com/mongodb/terraform-provider-mongodbatlas/pull/2519#discussion_r1728618273

**Review by oarbusi — APPROVED** [review]
LGTM

**Pull request merged**